### PR TITLE
EN-9839: esdt tokens new endpoints

### PR DESCRIPTION
**Pull request by bogdan-rosianu**
Added 2 endpoints:
- `/address/<address>/esdts-with-role/<role>` that returns all token identifiers of the tokens where the given address has the given role assigned
- `/address/<address>/owned-nfts` that returns all token identifiers of the NFTs/SFTs where the given address is the owner

### CODE REVIEW

**Review by sasurobert — APPROVED** [review]

**Review by sasurobert — DISMISSED** [review]

File: api/address/routes.go [hunk 1]
```diff
@@ -459,6 +462,49 @@ func GetESDTTokensWithRole(c *gin.Context) {
 	)
 }
 
+// GetOwnedNFTs returns the token identifiers of the tokens where a given address is the owner
+func GetOwnedNFTs(c *gin.Context) {
```
Comment (sasurobert) [inline api/address/routes.go:466]: the naming is wrong. GetNFTTokenIDsRegisteredByAddress

File: node/node.go [hunk 2]
```diff
@@ -619,6 +619,71 @@ func (n *Node) GetESDTData(address, tokenID string, nonce uint64) (*esdt.ESDigit
...
+			continue
+		}
+
+		err = n.internalMarshalizer.Unmarshal(esdtToken, value)
+		if err != nil {
+			log.Warn("cannot unmarshal", "token name", tokenName, "err", err)
+			continue
+		}
+
+		if !bytes.Equal(esdtToken.TokenType, []byte(core.FungibleESDT)) && bytes.Equal(esdtToken.OwnerAddress, addressBytes) {
```
Comment (sasurobert) [inline node/node.go:679]: this is copy paste code - a lot of duplicates here. Merge them please. And you can pass a selector function for this IF. All the other data is the same.

**Review by sasurobert — CHANGES_REQUESTED** [review]

File: api/address/routes.go (hunk 1 above)
Comment (bogdan-rosianu) [reply api/address/routes.go:466]: done

File: node/node.go (hunk 2 above)
Comment (bogdan-rosianu) [reply node/node.go:679]: done

File: node/node.go [hunk 3]
```diff
@@ -615,6 +628,103 @@ func (n *Node) GetESDTData(address, tokenID string, nonce uint64) (*esdt.ESDigit
...
+
+	ctx := context.Background()
+	chLeaves, err := userAccount.DataTrie().GetAllLeavesOnChannel(rootHash, ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	for leaf := range chLeaves {
+		tokenName := string(leaf.Key())
+		if !strings.Contains(tokenName, "-") {
```
Comment (raduchis) [inline node/node.go:664]: Is this the only way to detect an ESDT? Looks unreliable

**Review by raduchis — COMMENTED** [review]
Also unit tests do not pass on Teamcity

File: api/address/routes.go [hunk 4]
```diff
@@ -387,6 +394,117 @@ func GetESDTBalance(c *gin.Context) {
...
+		return
+	}
+
+	role := c.Param("role")
+	if role == "" {
+		c.JSON(
+			http.StatusBadRequest,
+			shared.GenericAPIResponse{
+				Data:  nil,
+				Error: fmt.Sprintf("%s: %s", errors.ErrGetESDTBalance.Error(), errors.ErrEmptyKey.Error()),
```
[the remark this answers is not in the document]
Comment (bogdan-rosianu) [reply api/address/routes.go:423]: no. fixed also in other cases

File: facade/nodeFacade_test.go [hunk 5]
```diff
@@ -752,6 +752,25 @@ func TestNodeFacade_GetAllIssuedESDTs(t *testing.T) {
 	assert.Equal(t, expectedValue, res)
 }
 
+func TestNodeFacade_GetESDTsWithRole(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (bogdan-rosianu) [reply facade/nodeFacade_test.go:755]: added

File: node/node.go (hunk 3 above)
Comment (bogdan-rosianu) [reply node/node.go:664]: Unfortunately, yes. Another approach would be to split the string by `-` and check that the second item is a 6 characters length. But, key-value pairs can only be added from the address and we don't have another key matching this pattern

File: api/address/routes.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (raduchis) [reply api/address/routes.go]: ok!

File: facade/nodeFacade_test.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (raduchis) [reply facade/nodeFacade_test.go:755]: ok!

File: node/node.go (hunk 3 above)
Comment (raduchis) [reply node/node.go:664]: o...k...

**Review by raduchis — APPROVED** [review]

**Review by gabi-vuls — APPROVED** [review]
System test passed

**Review by sasurobert — APPROVED** [review]

**Pull request merged**